client/shared/src/actions/ActionItem.tsx

```diff
@@ -356,6 +360,19 @@ export class ActionItem extends React.PureComponent<ActionItemProps, State, type
 
         // Record action ID (but not args, which might leak sensitive data).
         this.props.telemetryService.log(action.id)
+        if (action.telemetryProps) {
+            this.props.telemetryRecorder.recordEvent(
+                action.telemetryProps.feature as KnownString<Feature>,
```

Comment:
Here is the example I'm thinking of: https://sourcegraph.com/github.com/sourcegraph/cody/-/blob/agent/src/agent.ts?L684-695; this works because `'feature'` fulfills the type requirements of `KnownString` (sorry that my other comments on this were a bit confusing)

```suggestion
                action.telemetryProps.feature as 'feature',
```

Note the docstring I put above in particular that explains the usage:

```ts
// 👷 HACK: We have no control over what gets sent over JSON RPC,
                // so we depend on client implementations to give type guidance
                // to ensure that we don't accidentally share arbitrary,
                // potentially sensitive string values. In this RPC handler,
                // when passing the provided event to the TelemetryRecorder
                // implementation, we forcibly cast all the inputs below
                // (feature, action, parameters) into known types (strings
                // 'feature', 'action', 'key') so that the recorder will accept
                // it. DO NOT do this elsewhere!
```

I think we can adapt that docstring to this callsite as well

Reply:
No need to reply to this, the words being used here are hard to parse. I'll work on a next draft and get your take on that.